From ef6089b63cac2c6f92f00df82785f9d2f8aaf923 Mon Sep 17 00:00:00 2001
From: Aloka Dixit <quic_alokad@quicinc.com>
Date: Wed, 24 Jan 2024 13:50:54 -0800
Subject: [PATCH] wifi: cfg80211: allow background cac on device bandwidth

Allow userspace to start a background channel availability check (CAC)
on the current channel if device bandwidth is different than the
operating bandwidth. This allows radar detection on the portion of
the device bandwidth which is not part of the operating bandwidth.
The CAC duration will use the off-channel durations for the applicable
regulatory domain.
Add a new input parameter to cfg80211_chandef_dfs_cac_time() to indicate
if the CAC is required for the current device bandwidth to differentiate
it from the background CAC on a different channel.
Add checks for the device bandwidth parameter in
cfg80211_chandef_identical().

Signed-off-by: Aloka Dixit <quic_alokad@quicinc.com>
Signed-off-by: Shivani Tambatkar <quic_stambatk@quicinc.com>
---
 include/net/cfg80211.h | 24 ++++++++++++++---
 net/wireless/chan.c    | 61 +++++++++++++++++++++++++++++++++++++++++-
 net/wireless/core.h    |  3 ++-
 net/wireless/mlme.c    | 20 +++++++++++---
 net/wireless/nl80211.c | 17 ++++++++++--
 5 files changed, 114 insertions(+), 11 deletions(-)

diff --git a/include/net/cfg80211.h b/include/net/cfg80211.h
index d385dbc1b111..d7817aae6697 100644
--- a/include/net/cfg80211.h
+++ b/include/net/cfg80211.h
@@ -978,8 +978,9 @@ cfg80211_chandef_identical(const struct cfg80211_chan_def *chandef1,
 		chandef1->freq1_offset == chandef2->freq1_offset &&
 		chandef1->center_freq2 == chandef2->center_freq2 &&
 		chandef1->ru_punct_bitmap == chandef2->ru_punct_bitmap &&
-		chandef1->ru_punct_bitmap_supp_he ==
-					chandef2->ru_punct_bitmap_supp_he);
+		chandef1->ru_punct_bitmap_supp_he == chandef2->ru_punct_bitmap_supp_he &&
+		chandef1->width_device == chandef2->width_device &&
+		chandef1->center_freq_device == chandef2->center_freq_device);
 }
 
 /**
@@ -1159,17 +1160,32 @@ ieee80211_chandef_rate_flags(struct cfg80211_chan_def *chandef)
 bool cfg80211_chandef_dfs_usable(struct wiphy *wiphy,
 				 const struct cfg80211_chan_def *chandef);
 
+/**
+ * cfg80211_chandef_dfs_usable_device - checks if chandef is DFS usable
+ * @wiphy: the wiphy to validate against
+ * @chandef: the channel definition to check
+ *
+ * Checks if chandef is usable when the device bandwidth parameters are different
+ * than the operating bandwidth parameters and we can/need start CAC on such channel.
+ *
+ * Return: true if all channels available and at least
+ *	   one channel requires CAC (NL80211_DFS_USABLE)
+ */
+bool cfg80211_chandef_dfs_usable_device(struct wiphy *wiphy,
+					const struct cfg80211_chan_def *chandef);
+
 /**
  * cfg80211_chandef_dfs_cac_time - get maximum dfs cac time traversing
  * through the entire channel width
  * @wiphy: the wiphy to validate against
  * @chandef: the channel definition to check
  * @is_bgcac: background cac state
+ * @is_dbw_cac: cac on current device bandwidth
  * Returns: dfs cac time, or 0 is none found
  */
 unsigned int cfg80211_chandef_dfs_cac_time(struct wiphy *wiphy,
-                                           const struct cfg80211_chan_def *chandef,
-                                           bool is_bgcac);
+					   const struct cfg80211_chan_def *chandef,
+					   bool is_bgcac, bool is_dbw_cac);
 
 /**
  * ieee80211_chandef_max_power - maximum transmission power for the chandef
diff --git a/net/wireless/chan.c b/net/wireless/chan.c
index 2ffa11671796..f7fa434dd068 100644
--- a/net/wireless/chan.c
+++ b/net/wireless/chan.c
@@ -922,6 +922,29 @@ bool cfg80211_chandef_dfs_usable(struct wiphy *wiphy,
 }
 EXPORT_SYMBOL(cfg80211_chandef_dfs_usable);
 
+bool cfg80211_chandef_dfs_usable_device(struct wiphy *wiphy,
+					const struct cfg80211_chan_def *chandef)
+{
+	int width;
+	int r1;
+
+	if (WARN_ON(!cfg80211_chandef_valid(chandef)))
+		return false;
+
+	width = nl80211_chan_width_to_mhz(chandef->width_device);
+	if (width < 0)
+		return false;
+
+	r1 = cfg80211_get_chans_dfs_usable(wiphy,
+					   MHZ_TO_KHZ(chandef->center_freq_device),
+					   width, 0);
+
+	if (r1 > 0)
+		return true;
+
+	return false;
+}
+
 /*
  * Checks if center frequency of chan falls with in the bandwidth
  * range of chandef.
@@ -1247,9 +1270,40 @@ static unsigned int cfg80211_get_chans_dfs_cac_time(struct wiphy *wiphy,
 	return dfs_cac_ms;
 }
 
+static unsigned int cfg80211_get_chans_dfs_cac_time_dbw(struct wiphy *wiphy,
+							const struct cfg80211_chan_def *chandef)
+{
+	struct ieee80211_channel *c;
+	u32 freq, end_freq;
+	unsigned int dfs_cac_ms = 0;
+
+	end_freq = cfg80211_get_end_freq_device(chandef);
+	for (freq = cfg80211_get_start_freq_device(chandef);
+	     freq <= end_freq; freq += MHZ_TO_KHZ(20)) {
+		if (!cfg80211_is_freq_device_non_oper(chandef, freq))
+			continue;
+
+		c = ieee80211_get_channel_khz(wiphy, freq);
+		if (!c)
+			continue;
+
+		if (c->flags & IEEE80211_CHAN_DISABLED)
+			continue;
+
+		if (!(c->flags & IEEE80211_CHAN_RADAR))
+			continue;
+
+		if (c->dfs_cac_ms > dfs_cac_ms)
+			dfs_cac_ms = c->dfs_cac_ms;
+	}
+
+	return dfs_cac_ms;
+}
+
 unsigned int
 cfg80211_chandef_dfs_cac_time(struct wiphy *wiphy,
-			      const struct cfg80211_chan_def *chandef, bool is_bgcac)
+			      const struct cfg80211_chan_def *chandef, bool is_bgcac,
+			      bool is_dbw_cac)
 {
 	int width;
 	unsigned int t1 = 0, t2 = 0, dfs_cac_time;
@@ -1257,6 +1311,11 @@ cfg80211_chandef_dfs_cac_time(struct wiphy *wiphy,
 	if (WARN_ON(!cfg80211_chandef_valid(chandef)))
 		return 0;
 
+	if (is_dbw_cac) {
+		t1 = cfg80211_get_chans_dfs_cac_time_dbw(wiphy, chandef);
+		goto exit;
+	}
+
 	width = cfg80211_chandef_get_width(chandef);
 	if (width < 0)
 		return 0;
diff --git a/net/wireless/core.h b/net/wireless/core.h
index 970e668fb1eb..e09d6beccc60 100644
--- a/net/wireless/core.h
+++ b/net/wireless/core.h
@@ -485,7 +485,8 @@ void cfg80211_sched_dfs_chan_update(struct cfg80211_registered_device *rdev);
 int
 cfg80211_start_background_radar_detection(struct cfg80211_registered_device *rdev,
 					  struct wireless_dev *wdev,
-					  struct cfg80211_chan_def *chandef);
+					  struct cfg80211_chan_def *chandef,
+					  int link_id);
 
 void cfg80211_stop_background_radar_detection(struct wireless_dev *wdev);
 
diff --git a/net/wireless/mlme.c b/net/wireless/mlme.c
index 2d895c5276ab..fdfff092b6ed 100644
--- a/net/wireless/mlme.c
+++ b/net/wireless/mlme.c
@@ -1143,6 +1143,11 @@ __cfg80211_background_cac_event(struct cfg80211_registered_device *rdev,
 
 	netdev = wdev ? wdev->netdev : NULL;
 	nl80211_radar_notify(rdev, chandef, event, netdev, GFP_KERNEL);
+
+	if ((event == NL80211_RADAR_CAC_FINISHED || event == NL80211_RADAR_CAC_ABORTED) &&
+	    cfg80211_chandef_identical(wdev_chandef(wdev, 0), chandef)) {
+		rdev->background_radar_wdev = NULL;
+	}
 }
 
 static void
@@ -1188,8 +1193,10 @@ EXPORT_SYMBOL(cfg80211_background_cac_abort);
 int
 cfg80211_start_background_radar_detection(struct cfg80211_registered_device *rdev,
 					  struct wireless_dev *wdev,
-					  struct cfg80211_chan_def *chandef)
+					  struct cfg80211_chan_def *chandef,
+					  int link_id)
 {
+	struct cfg80211_chan_def *current_chandef;
 	unsigned int cac_time_ms;
 	int err;
 
@@ -1212,8 +1219,15 @@ cfg80211_start_background_radar_detection(struct cfg80211_registered_device *rde
 	if (err)
 		return err;
 
-	cac_time_ms = cfg80211_chandef_dfs_cac_time(&rdev->wiphy, chandef,
-						    true);
+	current_chandef = wdev_chandef(wdev, link_id);
+	if (current_chandef &&
+	    cfg80211_chandef_identical(current_chandef, chandef) &&
+	    cfg80211_chandef_device_present(chandef))
+		cac_time_ms = cfg80211_chandef_dfs_cac_time(&rdev->wiphy, chandef,
+							    true, true);
+	else
+		cac_time_ms = cfg80211_chandef_dfs_cac_time(&rdev->wiphy, chandef,
+							    true, false);
 	if (!cac_time_ms)
 		cac_time_ms = IEEE80211_DFS_MIN_CAC_TIME_MS;
 
diff --git a/net/wireless/nl80211.c b/net/wireless/nl80211.c
index 95c994c6c156..c36dbee334eb 100644
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -10880,6 +10880,19 @@ static int nl80211_start_radar_detection(struct sk_buff *skb,
 		goto unlock;
 	}
 
+	if (nla_get_flag(info->attrs[NL80211_ATTR_RADAR_BACKGROUND]) &&
+	    chandef_link && chandef_link->center_freq_device &&
+	    chandef_link->width_device &&
+	    cfg80211_chandef_identical(&chandef, chandef_link)) {
+		if (!cfg80211_chandef_dfs_usable_device(wiphy, &chandef)) {
+			err = -EINVAL;
+			goto unlock;
+		}
+		err = cfg80211_start_background_radar_detection(rdev, wdev,
+								&chandef, link_id);
+		goto unlock;
+	}
+
 	if (!cfg80211_chandef_dfs_usable(wiphy, &chandef)) {
 		err = -EINVAL;
 		goto unlock;
@@ -10887,7 +10900,7 @@ static int nl80211_start_radar_detection(struct sk_buff *skb,
 
 	if (nla_get_flag(info->attrs[NL80211_ATTR_RADAR_BACKGROUND])) {
 		err = cfg80211_start_background_radar_detection(rdev, wdev,
-								&chandef);
+								&chandef, link_id);
 		goto unlock;
 	}
 
@@ -10913,7 +10926,7 @@ static int nl80211_start_radar_detection(struct sk_buff *skb,
 	}
 
 	cac_time_ms = cfg80211_chandef_dfs_cac_time(&rdev->wiphy, &chandef,
-						    false);
+						    false, false);
 	if (WARN_ON(!cac_time_ms))
 		cac_time_ms = IEEE80211_DFS_MIN_CAC_TIME_MS;
 
-- 
2.17.1

